admin/services/pupil-frozen.service/pupil-frozen.data.service.ts

```diff
@@ -0,0 +1,80 @@
+const sqlService = require('../data-access/sql.service')
+import { TYPES } from '../data-access/sql.service'
+
+export class PupilFrozenDataService {
+/**
+ * @description look up a pupils frozen flag by their id
+ * @param {number} pupilId
+ * @returns {Promise<Array<any>>}
+ */
+  public static async isFrozen (pupilId: number): Promise<any> {
+    const sql = 'SELECT frozen FROM mtc_admin.pupil WHERE id=@pupilId'
+    const params = [{
+      name: 'pupilId',
+      type: TYPES.Int,
+      value: pupilId
+    }]
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+/**
+ * @description look up a pupils frozen flag by their urlSlug
+ * @param {string} pupilUrlSlug
+ * @returns {Promise<Array<any>>}
+ */
+  public static async isFrozenByUrlSlug (pupilUrlSlug: string): Promise<any> {
+    const sql = 'SELECT frozen FROM mtc_admin.pupil WHERE urlSlug=@pupilUrlSlug'
+    const params = [{
+      name: 'pupilUrlSlug',
+      type: TYPES.UniqueIdentifier,
+      value: pupilUrlSlug
+    }]
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+/**
+ * @description get the count of pupils who are frozen within a set of urlSlugs
+ * @param {Array<string>} pupilUrlSlugs
+ * @returns {Promise<Array<any>>}
+ */
+  public static async countFrozenByUrlSlugs (pupilUrlSlugs: Array<string>): Promise<any> {
+    const { params, paramIdentifiers } = sqlService.buildParameterList(pupilUrlSlugs, TYPES.UniqueIdentifier,
+      'pupilUrlSlug')
+    const sql = `SELECT COUNT(id) as [frozenCount]
+      FROM mtc_admin.pupil WHERE frozen=1 AND urlSlug IN (${paramIdentifiers.join(', ')})`
+    return sqlService.readonlyQuery(sql, params)
+  }
+
+  /**
+ * @description get the count of pupils who are frozen within a set of urlSlugs
+ * @param {Array<string>} pupilIds
+ * @returns {Promise<Array<any>>}
+ */
+   public static async countFrozenByPupilIds (pupilIds: Array<number>): Promise<any> {
+    const paramData = sqlService.buildParameterList(pupilIds, TYPES.Int,
+      'pupilId')
+    const sql = `SELECT COUNT(id) as [frozenCount]
+      FROM mtc_admin.pupil WHERE frozen=1 AND id IN (${paramData.paramIdentifiers.join(', ')})`
+    return sqlService.readonlyQuery(sql, paramData.params)
+  }
+
+  public static async freezePupil (pupilUrlSlug: string): Promise<void> {
```

Comment:
Why would we change the modify function? The return statement is in this method.

Reply:
overall, the comments in this PR are more about the fundamentals of async/await, returning work and what the outcome of that work is.  I think we need to clarify on best practice - should you avoid holding on for the outcome of a piece of work if you don't need to do anything with the return value? or does the high level promise pattern just 'deal with that under the hood'?

there are some subtle semantics in other languages for this particular scenario, but i would rather we get straight on best practices for typescript and keep it that way.

ultimately, my concern is that you want to hang about here (via the `await`) for a piece of work to finish, and then just return void, which seems like a waste of an asynchronous piece of work, because you are essentially blocking until its finished, then returning nothing.  what if the caller wanted the piece of work, then to wait until many of them were complete and do something with the aggregated outcome?